torch/testing/_internal/common_distributed.py

```diff
@@ -834,3 +839,69 @@ def tp_transports():
     see https://github.com/pytorch/pytorch/issues/73885 and https://github.com/pytorch/pytorch/issues/65022
     """
     return ["shm", "uv"] if has_efa() else None
+
+
+
+def spawn_threads_and_init_comms(func=None, timeout=TIMEOUT_DEFAULT, world_size=DEFAULT_WORLD_SIZE):
+    """
+    Wrapper to use with a test method
+    """
+    if func is None:
+        return partial(spawn_threads_and_init_comms, timeout=timeout, world_size=world_size)
+
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        failed_ranks = run_with_threaded_pg(world_size, timeout, lambda: func(self, *args, **kwargs))
+        for rank, exc_info in failed_ranks:
+            print(f"Rank {rank} raised:")
+            for line in traceback.format_exception(*exc_info):
+                sys.stdout.write(line)
+        self.assertEqual([], failed_ranks, "Some ranks failed")
+
+
+    return wrapper
+
+class MultiThreadedTestCase(TestCase):
+    """
+    Simple test runner that executes all tests with the in-proc process group.
+
+    A single instance of the TestCase object for all threads.
+
+    Difference from regular test runner:
+    Cannot use setUp / tearDown (must use perThreadSetup / perThreadShutdown)
+        Not sure what these two would be good for though.
+    No global state possible
+        How bad of a limitation is this?
```

Comment:
I think this should be okay, it sounds risky for unittests to rely on global state in general?

Reply:
Yes, it's bad for unit tests, but it might affect PT code that depends on it.
We'll likely run into it and have to deal case-by-case.